src/graph/optimizer/OptRule.cpp

```diff
@@ -76,6 +86,48 @@ StatusOr<MatchedResult> Pattern::match(const OptGroup *group) const {
   return Status::Error();
 }
 
+bool OptRule::TransformResult::checkDataFlow(const std::vector<OptGroup *> &boundary) {
+  return std::all_of(
+      newGroupNodes.begin(), newGroupNodes.end(), [&boundary](const OptGroupNode *groupNode) {
+        return checkDataFlow(groupNode, boundary);
+      });
+}
+
+/*static*/ bool OptRule::TransformResult::checkDataFlow(const OptGroupNode *groupNode,
+                                                        const std::vector<OptGroup *> &boundary) {
+  const auto &deps = groupNode->dependencies();
+  // reach the boundary
+  if (std::all_of(deps.begin(), deps.end(), [&boundary](OptGroup *dep) {
+        return std::find(boundary.begin(), boundary.end(), dep) != boundary.end();
+      })) {
+    return true;
+  }
+  const auto *group = groupNode->group();
+  if (std::find(boundary.begin(), boundary.end(), group) != boundary.end()) {
+    return true;
+  }
+  // Check dataflow
+  const auto *node = groupNode->node();
+  if (node->inputVars().size() == deps.size()) {
+    // Don't check when count of dependencies is different from count of input variables
+    for (std::size_t i = 0; i < deps.size(); i++) {
+      const OptGroup *dep = deps[i];
+      if (node->inputVar(i) != dep->outputVar()) {
+        return false;
+      }
+      // Only use by father plan node
+      if (node->inputVars()[i]->readBy.size() != 1) {
```

Comment:
it's better to check whether the input variables are from outside plan. But it's ok to process like this now.